### ISSUE TITLE
overlap of p5.Image._modified and p5.Image_pixelsDirty

### ISSUE THREAD

**Issue by lmccart** [processing/p5.js#3726]
#### Nature of issue?

- [ ] Found a bug
- [ x ] Existing feature enhancement
- [ ] New feature request

#### Most appropriate sub-area of p5.js?
- [ x ] Image

#### Feature enhancement details:

The p5.Image object has two properties [`this._modified` and `this._pixelsDirty`](https://github.com/processing/p5.js/blob/master/src/image/p5.Image.js#L151). They seem to overlap conceptually but are currently serving different purposes. It would be great to combine these if possible, or add inline comments that differentiate the two.

A note from @stalgiag:
With a bit of investigation it seems that _modified is only ever used by this one function (https://github.com/processing/p5.js/blob/master/src/webgl/p5.Texture.js#L138) but at first glance it might be kind of difficult to merge them as _modified needs to changed within that function while _pixelsDirty is used everywhere. That said, they definitely can be combined with some untangling and probably should be because it is a bit silly to have two booleans doing two conceptually similar things.

**Comment by Spongman** [comment]
I don’t think these two can be simply merged. One represents the state of the `_pixels` array, and the other handles the state of the texture on the GPU. Merging these would require updating the GPU every time the pixels array is synced which could be a big perf hit.

**Comment by stalgiag** [comment]
Good point @Spongman. I don't fully understand why it would be a major performance hit if the GPU texture was updated when `_pixelsDirty` is true. You did most of the work managing the `_pixelsDirty` flag, could you point to when `_pixelsDirty` would be true too often or when it would force an unnecessary update to the texture on the GPU? I am having a bit of trouble tracing this around but my understanding is that -
`p5.Image` has itself as `_pixelsState` and `_pixelsDirty` is only set to true when there is some modification to the pixels such as with `set()`.
If this understanding is correct, wouldn't we want to update the texture state on the GPU when this happens?
Alternatively if it isn't possible to use `_pixelsDirty` to determine when to update the texture on the GPU would it be possible to store `_modified` on the `p5.Texture` object so that its purpose is more clear? I understand this requires a bit of weirdness because many of the methods that modify won't know whether the image is part of a texture object. It is possible that a comment or a renaming is enough. Just curious to understand a bit better.

**Comment by stalgiag** [comment]
So I spent a bit of time testing things out and if `_pixelsDirty` is used to determine when to upload to the GPU , the number of times that `gl.texImage2D()` is the same for all the manual test examples. The problems come with needing to alter `_pixelsDirty` after using it inside of the texture update method. This works the same as well with all of our test examples but once I started playing around with adding `loadPixels()`, `set()`, and `updatePixels()` with different timing, I ran into problems with [updatePixels()](https://github.com/processing/p5.js/blob/master/src/core/p5.Renderer2D.js#L396) as in some cases `pixelsState.imageData` or `p5.Image.pixelsData` [hasn't been created](https://github.com/processing/p5.js/blob/master/src/webgl/p5.RendererGL.js#L637) because the `_pixelsDirty` flag had been modified by texture updating.

Anyways, I still think that there would be a way for these two booleans to track and manage state for the same purpose but I also can't fully follow the state management for `_pixelsDirty` and my suspicion is that it would take too much work to restructure the timing of state change for `_pixelsDirty` to make it serve both purposes. I am going to open a PR that offers a simple comment change.
